packages/flutter/lib/src/widgets/layout_builder.dart

```diff
@@ -0,0 +1,192 @@
+// Copyright 2016 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'debug.dart';
+import 'framework.dart';
+
+import 'package:flutter/rendering.dart';
+
+/// The signature of the [LayoutBuilder] builder function.
+typedef Widget LayoutWidgetBuilder(BuildContext context, Size size);
+
+/// Builds a widget tree that can depend on the parent widget's size.
+///
+/// Similar to the [Builder] widget except that the framework calls the [builder]
+/// function at layout time and provides the parent widget's size. This is useful
+/// when the parent constrains the child's size and doesn't depend on the child's
+/// intrinsic size.
+class LayoutBuilder extends RenderObjectWidget {
+  LayoutBuilder({ Key key, this.builder }) : super(key: key);
+
+  /// Called at layout time to construct the widget tree. The builder must not
+  /// return null.
+  final LayoutWidgetBuilder builder;
+
+  @override
+  _LayoutBuilderElement createElement() => new _LayoutBuilderElement(this);
+
+  @override
+  _RenderLayoutBuilder createRenderObject(BuildContext context) => new _RenderLayoutBuilder();
+}
+
+class _RenderLayoutBuilder extends RenderBox with RenderObjectWithChildMixin<RenderBox> {
+  _RenderLayoutBuilder({ LayoutCallback callback }) : _callback = callback;
+
+  LayoutCallback get callback => _callback;
+  LayoutCallback _callback;
+  void set callback(LayoutCallback value) {
+    if (value == _callback)
+      return;
+    _callback = value;
+    markNeedsLayout();
+  }
+
+  double getIntrinsicWidth(BoxConstraints constraints) => constraints.constrainWidth();
+
+  double getIntrinsicHeight(BoxConstraints constraints) => constraints.constrainHeight();
+
+  @override
+  double getMinIntrinsicWidth(BoxConstraints constraints) {
+    assert(constraints.debugAssertIsValid());
+    return getIntrinsicWidth(constraints);
+  }
+
+  @override
+  double getMaxIntrinsicWidth(BoxConstraints constraints) {
+    assert(constraints.debugAssertIsValid());
+    return getIntrinsicWidth(constraints);
+  }
+
+  @override
+  double getMinIntrinsicHeight(BoxConstraints constraints) {
+    assert(constraints.debugAssertIsValid());
+    return getIntrinsicHeight(constraints);
+  }
+
+  @override
+  double getMaxIntrinsicHeight(BoxConstraints constraints) {
+    assert(constraints.debugAssertIsValid());
+    return getIntrinsicHeight(constraints);
+  }
+
+  @override
+  bool get sizedByParent => true;
+
+  @override
+  void performResize() {
+    size = constraints.biggest;
+  }
+
+  @override
+  void performLayout() {
+    if (callback != null)
+      invokeLayoutCallback(callback);
+    if (child != null)
+      child.layout(constraints.loosen(), parentUsesSize: false);
```

Comment:
That's so that the size of the child that's built can be smaller than the parent.